src/gt4py/gtc/passes/gtir_dtype_resolver.py

```diff
@@ -0,0 +1,83 @@
+from eve import NodeTranslator
+
+from gt4py.gtc import gtir
+from gt4py.gtc.common import DataType
+
+from devtools import debug
+
+
+class _GTIRResolveAuto(NodeTranslator):
+    """
+    Replaces AUTO dtype by a concrete dtype.
+
+    Precondition: All dtype are set (not None)
+    Postcondition: All dtypes are concrete (no AUTO)
+    """
+
+    class _GTIRUpdateAutoDecl(NodeTranslator):
+        """Updates FieldDecls with resolved types"""
+
+        def visit_FieldDecl(self, node: gtir.FieldDecl, new_symbols, **kwargs):
+            if node.dtype == DataType.AUTO:
+                dtype = new_symbols[node.name].dtype
+                return gtir.FieldDecl(name=node.name, dtype=dtype)
+            else:
+                return node
+
+    def visit_FieldAccess(self, node: gtir.FieldAccess, *, symtable, **kwargs):
```

Comment:
TODO add comment that the only case is for temporaries (and therefore currently ScalarAccess is not processed)